mixnet/client/src/sender.rs

```diff
@@ -133,21 +133,60 @@ where
             return None;
         }
 
+        // corner case: 0 average delay
+        if self.avg_delay == 0 {
+            self.remaining_delays -= 1;
+            return Some(0);
+        }
+
         if self.remaining_delays == 1 {
             self.remaining_delays -= 1;
             return Some(self.remaining_time);
         }
 
+        // if we have no remaining time, then just return 0
+        if self.remaining_time == 0 {
+            self.remaining_delays -= 1;
+            return Some(0);
+        }
+
         // Calculate bounds to avoid extreme values
         let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0)
             as u64;
```

Comment:
It would be helpful if we add a comment that explains why we need this `min(...)`. I understood it's for guaranteeing at least 1ms delay for all remaining hops. Did I understand it correctly?

Reply:
Sorry for missing the comment, and yes, this is a guarantee of at least a 1ms delay for all remaining hops.